rewriter/RBSAssertions.cc

```diff
@@ -0,0 +1,509 @@
+     *       @x = T.let(@x, T.nilable(Integer))
+     *       <tmp> = 42
+     *       @x = <tmp>
+     *     end
+     *
+     * Since with RBS comments we introduce the `T.let` after the desugar we need to manually emulate this
+     * behavior and transform the `else` branch accordingly.
+     *
+     * Will apply the transformation if and only if:
+     *
+     *  1. The condition is an `UnresolvedIdent` and it's an instance variable
+     *  2. The `then` branch returns the same instance variable
+     *  3. The `else` branch is an `Assign` to the same instance variable using a `T.let` of a nilable type
+     */
+    void postTransformIf(core::MutableContext ctx, ast::ExpressionPtr &tree) {
+        auto iff = ast::cast_tree<ast::If>(tree);
+
+        // Is the condition an `UnresolvedIdent` and it's an instance variable?
+        auto condIdent = ast::cast_tree<ast::UnresolvedIdent>(iff->cond);
+        if (!condIdent || condIdent->kind != ast::UnresolvedIdent::Kind::Instance) {
+            return;
+        }
+
+        // Is the `then` branch an `UnresolvedIdent` and it's the same instance variable?
+        auto name = condIdent->name;
+        auto thenIdent = ast::cast_tree<ast::UnresolvedIdent>(iff->thenp);
+        if (!thenIdent || thenIdent->name != name) {
+            return;
+        }
+
+        // Is the `else` branch an `Assign` to the same instance variable using a `T.let` of a nilable type?
+        auto elseAssign = ast::cast_tree<ast::Assign>(iff->elsep);
+        if (!elseAssign) {
+            return;
+        }
+        auto elseAssignLhs = ast::cast_tree<ast::UnresolvedIdent>(elseAssign->lhs);
+        if (!elseAssignLhs || elseAssignLhs->name != name) {
+            return;
+        }
+        auto elseAssignCast = ast::cast_tree<ast::Cast>(elseAssign->rhs);
+        if (!elseAssignCast) {
+            return;
+        }
+        if (!ast::MK::isTNilable(elseAssignCast->typeExpr)) {
+            return;
+        }
+
+        // We have a else branch that looks like this:
+        //
+        //     @x = T.let(42, T.nilable(Integer))
+        //
+        // We need to transform it into:
+        //
+        //     @x = T.let(@x, T.nilable(Integer))
+        //     <tmp> = 42
+        //     @x = <tmp>
+
+        auto stats = ast::InsSeq::STATS_store();
+        stats.reserve(3);
+
+        auto let = ast::MK::Let(elseAssignCast->loc, ast::MK::cpRef(elseAssign->lhs), move(elseAssignCast->typeExpr));
+        auto newLet = ast::make_expression<ast::Assign>(elseAssign->loc, ast::MK::cpRef(elseAssign->lhs), move(let));
+        stats.push_back(move(newLet));
+
+        auto tempLocal = ast::MK::Local(elseAssignCast->loc, core::Names::statTemp());
+        auto newTemp = ast::MK::Assign(elseAssignCast->loc, ast::MK::cpRef(tempLocal), move(elseAssignCast->arg));
+        stats.push_back(move(newTemp));
+
+        auto expr = ast::make_expression<ast::Assign>(elseAssign->loc, move(elseAssign->lhs), move(tempLocal));
```

Comment:
Why are we not using `MK::Assign`? I believe that function is intending to preserve some invariants that should be established universally after desugar runs.